package/MDAnalysis/topology/LAMMPSParser.py

```diff
@@ -170,59 +217,34 @@ def read_DATA_timestep(self, n_atoms, TS_class, TS_kwargs):
 
         .. versionadded:: 0.9.0
         """
-        read_coords = False
-        read_velocities = False
-        velocities = None
-
-        with openany(self.filename, 'r') as datafile:
-            nitems, ntypes, box = self._parse_header(datafile)
-
-            # lammps box: xlo, xhi, ylo, yhi, zlo, zhi
-            lx = box[1] - box[0]
-            ly = box[3] - box[2]
-            lz = box[5] - box[4]
-            # mda unitcell: A alpha B beta gamma C
-            unitcell = np.zeros(6, dtype=np.float32)
-            unitcell[[0, 1, 2]] = lx, ly, lz
-            unitcell[[3, 4, 5]] = 90.0
+        header, sects = self.grab_datafile()
 
-            while not (read_coords & read_velocities):
-                try:
-                    section = datafile.next().strip().split()[0]
-                except IndexError:  # blank lines don't split
-                    section = ''
-                except StopIteration:
-                    break
-
-                if section == 'Atoms':
-                    positions = np.zeros((n_atoms, 3),
-                                         dtype=np.float32, order='F')
-                    self._parse_pos(datafile, positions)
-                    read_coords = True
-                elif section == 'Velocities':
-                    velocities = np.zeros((n_atoms, 3),
-                                          dtype=np.float32, order='F')
-                    self._parse_vel(datafile, velocities)
-                    read_velocities = True
-                elif len(section) > 0:
-                    self._skip_section(datafile)
-                else:
-                    continue
+        unitcell = self._parse_box(header)
 
-        if not read_coords:  # Reaches here if StopIteration hit
+        positions = np.zeros((n_atoms, 3),
+                             dtype=np.float32, order='F')
```

Comment:
On 19 Nov, 2015, at 12:08, Hai Nguyen wrote:

> @richardjgowers copying data is very expensive. Can mda avoid it?

Generally speaking, avoiding copies would be nice, also in the context of e.g. distance_array() and friends.

IIRC at the dawn of MDA we might not have had the right tools to make this work cleanly (or simply not the experience); perhaps with cleaner numpy/cython interfaces, memviews etc this might be easier nowadays.

It would be useful to have a separate issue that simply lists the places where we could improve the code in this manner.